src/charm_state.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""This module defines the CharmState class which represents the state of the Flask charm."""
+
+import datetime
+import pathlib
+
+from ops.charm import CharmBase
+
+from charm_types import WebserverConfig
+
+
+class CharmState:
+    """Represents the state of the Flask charm."""
+
+    def __init__(
+        self,
+        webserver_workers: int | None = None,
+        webserver_threads: int | None = None,
+        webserver_keepalive: int | None = None,
+        webserver_timeout: int | None = None,
+    ):
+        """Initialize a new instance of the CharmState class.
+
+        Args:
+            webserver_workers: The number of workers to use for the web server,
+                or None if not specified.
+            webserver_threads: The number of threads per worker to use for the web server,
+                or None if not specified.
+            webserver_keepalive: The time to wait for requests on a Keep-Alive connection,
+                or None if not specified.
+            webserver_timeout: The request silence timeout for the web server,
+                or None if not specified.
+        """
+        self._webserver_workers = webserver_workers
+        self._webserver_threads = webserver_threads
+        self._webserver_keepalive = webserver_keepalive
+        self._webserver_timeout = webserver_timeout
+
+    @classmethod
+    def from_charm(cls, charm: CharmBase) -> "CharmState":
+        """Initialize a new instance of the CharmState class from the associated charm.
+
+        Args:
+            charm: The charm instance associated with this state.
+
+        Return:
+            The CharmState instance created by the provided charm.
+        """
+        keepalive = charm.config.get("webserver_keepalive")
+        timeout = charm.config.get("webserver_timeout")
+        workers = charm.config.get("webserver_workers")
+        threads = charm.config.get("webserver_threads")
+        return cls(
+            webserver_workers=int(workers) if workers is not None else None,
```

Comment:
What is the reason for the `is None` check please? It seems that `__init__` is fine with taking `None` values?

Reply:
This is just for type checks since the `charm.config` is typed with `Map[str, str | None]`

Reply:
Potentially worth adding a comment to explain, up to you